M2/Macaulay2/packages/GKMManifolds.m2

```diff
@@ -0,0 +1,1468 @@
+net MomentGraph := G -> net ofClass class G | " on " | toString(#G.vertices) | " vertices with " | toString(#G.edges) | " edges "
+
+--given a list V representing the vertices, E a hash table of edges, and H a ring for HTpt
+--outputs a moment graph with such data
+momentGraph = method()
+momentGraph(List,HashTable,Ring) := MomentGraph => (V,E,H) -> (
+    new MomentGraph from {
+	symbol vertices => V,
+	symbol edges => E,
+	symbol HTpt => H,
+	cache => new CacheTable
+	}
+    )
+
+
+--outputs the graph underlying a moment graph G
+underlyingGraph(MomentGraph) := Graph => G -> graph(G.vertices, keys G.edges, EntryMode => "edges")
+
+
+--If a moment graph G comes from a possibly singular GKM variety with a T-invariant
+-- Whitney stratification consisting of affine spaces,
+--the vertices of G correspond to each strata,
+--and are ordered v1 <= v2 if the closure of stratum of v2 is contained in that of v1.
+cellOrder = method()
+cellOrder(MomentGraph,Poset) := (G,P) -> (
+    if G.cache.?cellOrder then (
+	print "warning: overwriting a previously defined cell order on this moment graph "
+	);
+    if G.vertices =!= P.GroundSet then (
+	<< "the ground set of the poset is not the vertices of this moment graph " <<
+	return error
+	);
+    G.cache.cellOrder = P;
+    )
+
+
+cellOrder(MomentGraph) := Poset => G -> (
+    if G.cache.?cellOrder then G.cache.cellOrder
+    else << " no cell order defined on this moment graph" << return error
+    )
+
+--------------------------------------< T-varieties >-------------------------------------------
+
+
+--for setting up a character ring: given n, outputs the character ring of T = (kk^*)^n
+makeCharRing = method();
+makeCharRing(ZZ) := Ring => n -> (
+    T := symbol T;
+    ZZ[T_0..T_(n-1), MonomialOrder => GLex, Inverses=>true]
+)
+
+
+--a TVariety X is a HashTable with minimal data of:
+--X.points : a list representing torus-invariant points
+--X.charRing : a ring representing the character ring of the torus
+--additionally, a TVariety X can have:
+--X.charts : a hash table whose keys are X.points and values are (negative of) the characters 
+--of the affine chart at the point
+--X.momentGraph : a MomentGraph which is the moment graph of the T-variety
+TVariety = new Type of MutableHashTable
+
+TVariety.synonym = "T-variety"
+
+globalAssignment TVariety
+net TVariety := X -> net ofClass class X | " with an action of a " | toString(#gens X.charRing) | "-dimensional torus"
+
+
+tVariety = method()
+
+--a list L representing the torus-fixed points, and a character ring R defines a TVariety
+tVariety(List, Ring) := TVariety => (L,R) -> (
+    new TVariety from {
+	symbol points => L,
+	symbol charRing => R,
+	cache => new CacheTable
+	}
+    )
+
+
+--tVariety created from G, a moment graph, and a character ring R compatible with G.HTpt
+tVariety(MomentGraph,Ring) := TVariety => (G,R) -> (
+    if not #(gens R) == #(gens G.HTpt) then (
+	<< "HTpt not compatible with the character ring" <<
+	return error
```

Comment:
Probably you want to replace these two lines by `error "HTpt not compatible with the character ring"`.

Reply:
done